File: PrimaMateria/PrimaMateria/Classes/XTRWikipediaViewController.m
//
//  XTRWikipediaViewController.m
//  PrimaMateria
//
//  Created by Jerry Porter on 11/21/11.
//  Copyright 2014 xTrensa. All rights reserved.
//

@interface XTRWikipediaViewController ()

- (void)prepareRequest;

@end

@implementation XTRWikipediaViewController

#pragma mark Private Methods

- (void)prepareRequest {
    NSURL *url = [NSURL URLWithString: [NSString stringWithFormat: @"https://en.wikipedia.org/wiki/%@", self.elementName]];
    NSMutableURLRequest *requestObj = [NSMutableURLRequest requestWithURL: url];
    
    [self.webView loadRequest: requestObj];
}

#pragma mark - Action Methods

- (IBAction)dismiss: (id)sender {
    [self dismissViewControllerAnimated: YES completion:nil];
}

-(IBAction)backButtonPressed: (id)sender {
    [self.progressHUD showAnimated: YES];
    [self.webView goBack];
}

-(IBAction)forwardButtonPressed: (id)sender {
    [self.progressHUD showAnimated: YES];
    [self.webView goForward];
}

#pragma mark - WebView Delegate Methods

- (void)webViewDidStartLoad: (UIWebView *)aWebView {
    [self.progressHUD showAnimated: YES];
    
    self.backButton.enabled = NO;
    self.forwardButton.enabled = NO;
    self.backButton.tintColor = UIColor.blackColor;
    self.forwardButton.tintColor = UIColor.blackColor;
}

- (void)webViewDidFinishLoad: (UIWebView *)aWebView {
    [MBProgressHUD hideHUDForView: self.view animated: YES];
    
    if(aWebView.canGoBack) {
        self.backButton.enabled = YES;
        self.backButton.tintColor = UIColor.whiteColor;
        self.forwardButton.tintColor = UIColor.blackColor;
    } else if(aWebView.canGoForward) {
        self.forwardButton.enabled = YES;
        self.backButton.tintColor = UIColor.blackColor;
        self.forwardButton.tintColor = UIColor.whiteColor;
    }
}

- (void)webView: (UIWebView *)webView didFailLoadWithError: (nullable NSError *)error {
    NSURL *url = [NSURL fileURLWithPath: [[NSBundle mainBundle] pathForResource: @"LoadFailure" ofType: @"html"]];
    NSMutableURLRequest *requestObj = [NSMutableURLRequest requestWithURL: url];
    
    [MBProgressHUD hideHUDForView: self.view animated: YES];
    
    [self.webView loadRequest: requestObj];
}

#pragma mark - View Management Methods

-(void)viewDidLoad {
    [super viewDidLoad];
    NSDictionary *textAttributes = @{NSForegroundColorAttributeName: UIColor.whiteColor, NSFontAttributeName: [UIFont boldSystemFontOfSize: 20]};
    [self.titleButtonItem setTitleTextAttributes: textAttributes forState: UIControlStateNormal];
    
    self.preferredContentSize = CGSizeMake(768, 620);

    self.progressHUD = [[MBProgressHUD alloc] initWithView: self.view];
    self.progressHUD.label.font = [UIFont fontWithName: @"Verdana-Bold" size: 26.0];
    self.progressHUD.detailsLabel.font = [UIFont fontWithName: @"Verdana-Bold" size: 15.0];
    self.progressHUD.delegate = self;
    self.progressHUD.label.text = @"Please Wait";
    self.progressHUD.detailsLabel.text = [NSString stringWithFormat: @"Loading Wikipedia Page for element: %@.", self.elementName];
}

- (void)viewWillAppear: (BOOL) animated {
    [super viewWillAppear:animated];
    
    self.titleButtonItem.title = [NSString stringWithFormat: @"Wikipedia Entry for element:  %@", self.elementName];
    
    self.backButton.enabled = NO;
    self.backButton.tintColor = UIColor.blackColor;
    self.forwardButton.enabled = NO;
    self.forwardButton.tintColor = UIColor.blackColor;
        
    [self.view addSubview: self.progressHUD];
    [self.progressHUD showAnimated: YES];
    [self prepareRequest];
}

- (BOOL)shouldAutorotateToInterfaceOrientation: (UIInterfaceOrientation) interfaceOrientation {
    return UIInterfaceOrientationIsLandscape(interfaceOrientation);
}

#pragma mark - MBProgressHUDDelegate methods

- (void)hudWasHidden: (MBProgressHUD *)aProgressHUD {
    [aProgressHUD removeFromSuperview];
}

#pragma mark - Memory Management Methods

- (void)dealloc {
    self.backButton = nil;
    self.forwardButton = nil;
    self.titleButtonItem = nil;
    self.webView.delegate = nil;
    self.webView = nil;
}

@end

File: PrimaMateria/PrimaMateria/Classes/XTRWikipediaViewController.swift
//
//  XTRWikipediaViewController.swift
//  PrimaMateria
//
//  Created by Jerry Porter on 4/14/16.
//  Copyright ©2018 xTrensa. All rights reserved.
//

import WebKit

class XTRWikipediaViewController : UIViewController, WKNavigationDelegate, MBProgressHUDDelegate {
    
    @IBOutlet var backButton : UIBarButtonItem!
    @IBOutlet var titleButtonItem: UIBarButtonItem!
    @IBOutlet var forwardButton : UIBarButtonItem!
    @IBOutlet var webView : WKWebView!
    
    private var progressHUD : MBProgressHUD?
    private var responseData : NSMutableData?
    private var request : NSMutableURLRequest?
    
    var elementName : String?
    
    // MARK: - Initialization Methods
    
    required init?(coder aDecoder: NSCoder) {
        super.init(coder: aDecoder)!
    }
    
    // MARK: - Internal Methods
    
    func prepareRequest() {        
        webView.load(URLRequest(url: URL(string: "https://en.wikipedia.org/wiki/\(elementName!)")!))
    }
    
    // MARK: - Action Methods
    
    @IBAction func dismiss(_ sender: UIButton) {
        dismiss(animated: true, completion:  nil)
    }
    
    @IBAction func backButtonPressed(_ sender: UIButton) {
        progressHUD?.show(animated: true)
        webView.goBack()
    }
    
    @IBAction func forwardButtonPressed(_ sender: UIButton) {
        progressHUD?.show(animated: true)
        webView.goForward()
    }
    
    // MARK: - WebView Delegate Methods
    
    //func webViewDidStartLoad(_ webView: UIWebView) {
    func webView(_ webView: WKWebView, didCommit navigation: WKNavigation!) {
        progressHUD?.show(animated: true)
        backButton.isEnabled = false
        forwardButton.isEnabled = false
        backButton.tintColor = UIColor.black
        forwardButton.tintColor = UIColor.black
    }
    
    //func webViewDidFinishLoad(_ aWebView: UIWebView) {
    public func webView(_ aWebView: WKWebView, didFinish navigation: WKNavigation!) {
        MBProgressHUD.hide(for: view, animated: true)
        
        if aWebView.canGoBack {
            backButton.isEnabled = true
            backButton.tintColor = UIColor.white
        } else if !aWebView.canGoBack {
            backButton.isEnabled = false
            backButton.tintColor = UIColor.black
        }
        
        if aWebView.canGoForward {
            forwardButton.isEnabled = true
            forwardButton.tintColor = UIColor.white
        } else if !aWebView.canGoForward {
            forwardButton.isEnabled = false
            forwardButton.tintColor = UIColor.black
        }
    }
    
    //func webView(_ webView: UIWebView, didFailLoadWithError error: Error) {
    public func webView(_ webView: WKWebView, didFailProvisionalNavigation navigation: WKNavigation!, withError error: Error) {
        let path = Bundle.main.path(forResource: "LoadFailure", ofType: FileType.html)!

        MBProgressHUD.hide(for: view, animated: true)
        
        webView.load(URLRequest(url: URL(fileURLWithPath: path)))
    }
    
    // MARK: - MBProgressHUDDelegate methods
    
    func hudWasHidden(_ aProgressHUD: MBProgressHUD) {
        aProgressHUD.removeFromSuperview()
    }
    
    // MARK: - View Management Methods
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        let textAttributes: [NSAttributedStringKey: AnyObject]? = [
            NSAttributedStringKey.foregroundColor: UIColor.white,
            NSAttributedStringKey.font: UIFont.boldSystemFont(ofSize: 20)
        ]
        
        titleButtonItem.setTitleTextAttributes(textAttributes, for: UIControlState())
        
        preferredContentSize = CGSize(width: 768, height: 620)
        
        progressHUD = MBProgressHUD(view: view)
        progressHUD!.label.font = UIFont.boldSystemFont(ofSize: 26)
        progressHUD!.detailsLabel.font = UIFont.boldSystemFont(ofSize: 15)
        progressHUD!.delegate = self
        progressHUD!.label.text = "Please Wait"
        progressHUD!.detailsLabel.text = "Loading Wikipedia Page for element: \(elementName!)."
        webView.navigationDelegate = self
   }
    
    override func viewWillAppear(_ animated: Bool)  {
        super.viewWillAppear(animated)
        
        titleButtonItem.title = "Wikipedia Entry for element: \(elementName!)"
        backButton.isEnabled = false
        backButton.tintColor = UIColor.black
        forwardButton.isEnabled = false
        forwardButton.tintColor = UIColor.black
        
        view.addSubview(progressHUD!)
        progressHUD?.show(animated: true)
        prepareRequest()
    }
    
    override var shouldAutorotate : Bool {
        return false
    }
    
    override var supportedInterfaceOrientations : UIInterfaceOrientationMask {
        return .landscape
    }
    
    // MARK: - Memory Management Methods
    
    deinit {
        backButton = nil
        forwardButton = nil
        titleButtonItem = nil
        webView.navigationDelegate = nil
        webView = nil
    }
    
}
